File: php/src/Domains/Subscriptions.php
<?php

namespace SignalHouse\SDK\Domains;

use SignalHouse\SDK\HttpClient;

class Subscriptions
{
    private HttpClient $client;
    private bool $enableAdmin;
    public ?object $admin = null;

    public function __construct(HttpClient $client, bool $enableAdmin)
    {
        $this->client = $client;
        $this->enableAdmin = $enableAdmin;

        if ($enableAdmin) {
            $this->admin = new class($client) {
                private HttpClient $client;

                public function __construct(HttpClient $client)
                {
                    $this->client = $client;
                }

                /**
                 * Get subscription templates
                 */
                public function getTemplates(?string $templateId = null, array $options = []): array
                {
                    $queryString = $this->client->getQueryString(['templateId' => $templateId]);
                    return $this->client->request("/subscription/template{$queryString}", array_merge([
                        'method' => 'GET',
                    ], $options));
                }

                /**
                 * Create a new subscription template
                 */
                public function createTemplate(array $templateData, array $options = []): array
                {
                    $this->client->require(['templateData' => $templateData]);
                    return $this->client->request('/subscription/template', array_merge([
                        'method' => 'POST',
                        'body' => $templateData,
                    ], $options));
                }

                /**
                 * Update a subscription template
                 */
                public function updateTemplate(string $templateId, array $templateData, array $options = []): array
                {
                    $this->client->require(['templateId' => $templateId, 'templateData' => $templateData]);
                    $safeTemplateId = rawurlencode($templateId);
                    return $this->client->request("/subscription/template/{$safeTemplateId}", array_merge([
                        'method' => 'PUT',
                        'body' => $templateData,
                    ], $options));
                }

                /**
                 * Delete a subscription template
                 */
                public function deleteTemplate(string $templateId, array $options = []): array
                {
                    $this->client->require(['templateId' => $templateId]);
                    $safeTemplateId = rawurlencode($templateId);
                    return $this->client->request("/subscription/template/{$safeTemplateId}", array_merge([
                        'method' => 'DELETE',
                    ], $options));
                }

                /**
                 * Create a custom subscription for a group
                 */
                public function createCustomSubscription(string $groupId, array $subscriptionData, array $options = []): array
                {
                    $this->client->require(['groupId' => $groupId, 'subscriptionData' => $subscriptionData]);
                    $safeGroupId = rawurlencode($groupId);
                    return $this->client->request("/subscription/user/custom/{$safeGroupId}", array_merge([
                        'method' => 'POST',
                        'body' => $subscriptionData,
                    ], $options));
                }
            };
        }
    }

    /**
     * Get subscription history for a group
     *
     * @param string $groupId The group ID
     * @param bool|null $onlyActive Whether to only include active subscriptions
     * @param array $options Additional request options
     * @return array The response from the server
     */
    public function getSubscriptions(string $groupId, ?bool $onlyActive = null, array $options = []): array
    {
        $queryString = $this->client->getQueryString(['groupId' => $groupId, 'onlyActive' => $onlyActive]);
        return $this->client->request("/subscription/user{$queryString}", array_merge([
            'method' => 'GET',
        ], $options));
    }

    /**
     * Subscribe a group to a template
     *
     * @param string $groupId The group ID
     * @param string $templateId The template ID
     * @param array $options Additional request options
     * @return array The response from the server
     */
    public function subscribe(string $groupId, string $templateId, array $options = []): array
    {
        $this->client->require(['groupId' => $groupId, 'templateId' => $templateId]);
        $safeGroupId = rawurlencode($groupId);
        $safeTemplateId = rawurlencode($templateId);
        return $this->client->request("/subscription/user/subscribe/{$safeGroupId}/{$safeTemplateId}", array_merge([
            'method' => 'POST',
            'body' => (object) [],
        ], $options));
    }

    /**
     * Unsubscribe a group
     *
     * @param string $groupId The group ID
     * @param array $options Additional request options
     * @return array The response from the server
     */
    public function unsubscribe(string $groupId, array $options = []): array
    {
        $this->client->require(['groupId' => $groupId]);
        $safeGroupId = rawurlencode($groupId);
        return $this->client->request("/subscription/user/cancel/{$safeGroupId}", array_merge([
            'method' => 'POST',
        ], $options));
    }
}

File: python/signalhouse/domains/subscriptions.py
"""Subscriptions domain for the SignalHouse SDK."""

from __future__ import annotations

from typing import Any, TYPE_CHECKING
from urllib.parse import quote

if TYPE_CHECKING:
    from ..client import SignalHouseSDK


class SubscriptionsAdmin:
    """Admin-only subscription operations (SignalHouse staff)."""

    def __init__(self, sdk: SignalHouseSDK) -> None:
        self._sdk = sdk

    def get_templates(
        self,
        *,
        template_id: str | None = None,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Get subscription templates with optional filtering by template ID.

        Args:
            template_id: The ID of a specific template to retrieve.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.
        """
        query_string = self._sdk._get_query_string({"templateId": template_id})
        return self._sdk._request(
            f"/subscription/template{query_string}",
            method="GET",
            token=token,
            headers=headers,
        )

    def create_template(
        self,
        template_data: dict[str, Any],
        *,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Create a new subscription template.

        Args:
            template_data: The data for the new template, including templateName and monthlyFee.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If template_data is missing.
        """
        self._sdk._require({"templateData": template_data})
        return self._sdk._request(
            "/subscription/template",
            method="POST",
            body=template_data,
            token=token,
            headers=headers,
        )

    def update_template(
        self,
        template_id: str,
        template_data: dict[str, Any],
        *,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Update an existing subscription template.

        Args:
            template_id: The ID of the subscription template to update.
            template_data: The data for the template, including templateName and monthlyFee.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If template_id or template_data is missing.
        """
        self._sdk._require({"templateId": template_id, "templateData": template_data})
        safe_template_id = quote(str(template_id), safe="")
        return self._sdk._request(
            f"/subscription/template/{safe_template_id}",
            method="PUT",
            body=template_data,
            token=token,
            headers=headers,
        )

    def delete_template(
        self,
        template_id: str,
        *,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Delete an existing subscription template.

        Args:
            template_id: The ID of the subscription template to delete.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If template_id is missing.
        """
        self._sdk._require({"templateId": template_id})
        safe_template_id = quote(str(template_id), safe="")
        return self._sdk._request(
            f"/subscription/template/{safe_template_id}",
            method="DELETE",
            token=token,
            headers=headers,
        )

    def create_custom_subscription(
        self,
        group_id: str,
        subscription_data: dict[str, Any],
        *,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Create a custom subscription for a group.

        Args:
            group_id: The ID of the group to create the custom subscription for.
            subscription_data: The data for the custom subscription.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If group_id or subscription_data is missing.
        """
        self._sdk._require({"groupId": group_id, "subscriptionData": subscription_data})
        safe_group_id = quote(str(group_id), safe="")
        return self._sdk._request(
            f"/subscription/user/custom/{safe_group_id}",
            method="POST",
            body=subscription_data,
            token=token,
            headers=headers,
        )


class Subscriptions:
    """Subscription management operations."""

    def __init__(self, sdk: SignalHouseSDK) -> None:
        self._sdk = sdk
        self.admin: SubscriptionsAdmin | None = None
        if sdk.enable_admin:
            self.admin = SubscriptionsAdmin(sdk)

    def get_subscriptions(
        self,
        *,
        group_id: str | None = None,
        only_active: bool | None = None,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Get a list of subscription history for a group.

        Args:
            group_id: The ID of the group to get subscription history for.
            only_active: Whether to only include active subscriptions.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.
        """
        query_string = self._sdk._get_query_string({
            "groupId": group_id,
            "onlyActive": only_active,
        })
        return self._sdk._request(
            f"/subscription/user{query_string}",
            method="GET",
            token=token,
            headers=headers,
        )

    def subscribe(
        self,
        group_id: str,
        template_id: str,
        *,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Subscribe a group to a template.

        Args:
            group_id: The ID of the group to subscribe.
            template_id: The ID of the template to subscribe the group to.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If group_id or template_id is missing.
        """
        self._sdk._require({"groupId": group_id, "templateId": template_id})
        safe_group_id = quote(str(group_id), safe="")
        safe_template_id = quote(str(template_id), safe="")
        return self._sdk._request(
            f"/subscription/user/subscribe/{safe_group_id}/{safe_template_id}",
            method="POST",
            body={},
            token=token,
            headers=headers,
        )

    def unsubscribe(
        self,
        group_id: str,
        *,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Unsubscribe a group from a template.

        Args:
            group_id: The ID of the group to unsubscribe.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If group_id is missing.
        """
        self._sdk._require({"groupId": group_id})
        safe_group_id = quote(str(group_id), safe="")
        return self._sdk._request(
            f"/subscription/user/cancel/{safe_group_id}",
            method="POST",
            token=token,
            headers=headers,
        )
